File: MCPForUnity/Editor/Tools/ManageCanvas.cs
using System;
using System.Collections.Generic;
using System.Text;
using MCPForUnity.Editor.Helpers;
using Newtonsoft.Json.Linq;
using UnityEditor;
using UnityEngine;
using UnityEngine.UI;

namespace MCPForUnity.Editor.Tools
{
    /// <summary>
    /// Builds runtime uGUI hierarchies (Canvas / RectTransform / Image / TextMeshProUGUI) from a declarative spec.
    /// Distinct from <see cref="ManageUI"/>, which targets UI Toolkit (UXML/USS/VisualElement).
    /// </summary>
    [McpForUnityTool("manage_canvas", AutoRegister = false, Group = "ui")]
    public static class ManageCanvas
    {
        private static readonly Dictionary<string, (Vector2 min, Vector2 max, Vector2 pivot)> AnchorPresets = new(StringComparer.OrdinalIgnoreCase)
        {
            { "top-left",        (new Vector2(0, 1),    new Vector2(0, 1),    new Vector2(0, 1)) },
            { "top-center",      (new Vector2(0.5f, 1), new Vector2(0.5f, 1), new Vector2(0.5f, 1)) },
            { "top-right",       (new Vector2(1, 1),    new Vector2(1, 1),    new Vector2(1, 1)) },
            { "middle-left",     (new Vector2(0, 0.5f), new Vector2(0, 0.5f), new Vector2(0, 0.5f)) },
            { "middle-center",   (new Vector2(0.5f, 0.5f), new Vector2(0.5f, 0.5f), new Vector2(0.5f, 0.5f)) },
            { "center",          (new Vector2(0.5f, 0.5f), new Vector2(0.5f, 0.5f), new Vector2(0.5f, 0.5f)) },
            { "middle-right",    (new Vector2(1, 0.5f), new Vector2(1, 0.5f), new Vector2(1, 0.5f)) },
            { "bottom-left",     (new Vector2(0, 0),    new Vector2(0, 0),    new Vector2(0, 0)) },
            { "bottom-center",   (new Vector2(0.5f, 0), new Vector2(0.5f, 0), new Vector2(0.5f, 0)) },
            { "bottom-right",    (new Vector2(1, 0),    new Vector2(1, 0),    new Vector2(1, 0)) },
            { "stretch-top",     (new Vector2(0, 1),    new Vector2(1, 1),    new Vector2(0.5f, 1)) },
            { "stretch-bottom",  (new Vector2(0, 0),    new Vector2(1, 0),    new Vector2(0.5f, 0)) },
            { "stretch-left",    (new Vector2(0, 0),    new Vector2(0, 1),    new Vector2(0, 0.5f)) },
            { "stretch-right",   (new Vector2(1, 0),    new Vector2(1, 1),    new Vector2(1, 0.5f)) },
            { "stretch-horizontal", (new Vector2(0, 0.5f), new Vector2(1, 0.5f), new Vector2(0.5f, 0.5f)) },
            { "stretch-vertical",   (new Vector2(0.5f, 0), new Vector2(0.5f, 1), new Vector2(0.5f, 0.5f)) },
            { "stretch",         (new Vector2(0, 0),    new Vector2(1, 1),    new Vector2(0.5f, 0.5f)) },
            { "stretch-all",     (new Vector2(0, 0),    new Vector2(1, 1),    new Vector2(0.5f, 0.5f)) },
        };

        private static Type s_tmpType;
        private static Type TmpTextType => s_tmpType ??= Type.GetType("TMPro.TextMeshProUGUI, Unity.TextMeshPro");

        public static object HandleCommand(JObject @params)
        {
            if (@params == null)
                return new ErrorResponse("Parameters cannot be null.");

            string action = @params["action"]?.ToString()?.ToLowerInvariant();
            if (action != "build")
                return new ErrorResponse($"Unknown action: '{action}'. Valid actions: build.");

            JObject layout = @params["layout"] as JObject;
            if (layout == null)
                return new ErrorResponse("'layout' is required (object describing the hierarchy).");

            string parentRef = @params["parent"]?.ToString() ?? layout["parent"]?.ToString();
            GameObject parentGO = ResolveParent(parentRef, out string parentError);
            if (parentError != null)
                return new ErrorResponse(parentError);

            try
            {
                var nodes = new List<object>();
                GameObject root = BuildNode(layout, parentGO?.transform, nodes);
                Undo.RegisterCreatedObjectUndo(root, "Build UI Layout");
                Selection.activeGameObject = root;
                EditorUtility.SetDirty(root);

                return new SuccessResponse(
                    $"Built layout '{root.name}' with {nodes.Count} node(s).",
                    new
                    {
                        root = new
                        {
                            name = root.name,
                            instanceID = root.GetInstanceID(),
                            path = GetGameObjectPath(root)
                        },
                        nodes
                    });
            }
            catch (Exception e)
            {
                McpLog.Error($"[ManageCanvas] build failed: {e}");
                return new ErrorResponse($"Failed to build layout: {e.Message}");
            }
        }

        private static GameObject BuildNode(JObject spec, Transform parent, List<object> registry)
        {
            string name = spec["name"]?.ToString() ?? "UI Element";

            var go = new GameObject(name, typeof(RectTransform));
            if (parent != null)
                go.transform.SetParent(parent, false);

            ConfigureRectTransform(go.GetComponent<RectTransform>(), spec["rect"] as JObject);

            if (spec["image"] is JObject imageSpec)
                ConfigureImage(go, imageSpec);

            if (spec["text"] is JObject textSpec)
                ConfigureText(go, textSpec);

            registry.Add(new
            {
                name = go.name,
                instanceID = go.GetInstanceID(),
                path = GetGameObjectPath(go)
            });

            if (spec["children"] is JArray children)
            {
                foreach (var child in children)
                {
                    if (child is JObject childSpec)
                        BuildNode(childSpec, go.transform, registry);
                }
            }

            return go;
        }

        private static void ConfigureRectTransform(RectTransform rt, JObject spec)
        {
            if (spec == null) return;

            string anchor = spec["anchor"]?.ToString();
            if (!string.IsNullOrEmpty(anchor) && AnchorPresets.TryGetValue(anchor, out var preset))
            {
                rt.anchorMin = preset.min;
                rt.anchorMax = preset.max;
                rt.pivot = preset.pivot;
            }

            if (TryReadVector2(spec["anchor_min"], out var aMin)) rt.anchorMin = aMin;
            if (TryReadVector2(spec["anchor_max"], out var aMax)) rt.anchorMax = aMax;
            if (TryReadVector2(spec["pivot"], out var pv)) rt.pivot = pv;

            if (TryReadVector2(spec["anchored_position"], out var ap)) rt.anchoredPosition = ap;
            if (TryReadVector2(spec["size_delta"], out var sd)) rt.sizeDelta = sd;

            if (spec["rotation"] != null)
            {
                try { rt.localEulerAngles = new Vector3(0, 0, spec["rotation"].ToObject<float>()); }
                catch { }
            }

            if (TryReadVector2(spec["scale"], out var sc))
                rt.localScale = new Vector3(sc.x, sc.y, 1f);
        }

        private static void ConfigureImage(GameObject go, JObject spec)
        {
            var img = go.GetComponent<Image>() ?? go.AddComponent<Image>();

            string spritePath = spec["sprite"]?.ToString();
            if (!string.IsNullOrEmpty(spritePath))
            {
                var sprite = AssetDatabase.LoadAssetAtPath<Sprite>(spritePath);
                if (sprite != null) img.sprite = sprite;
                else McpLog.Warn($"[ManageCanvas] Sprite not found at '{spritePath}' (on '{go.name}').");
            }

            if (TryReadColor(spec["color"], out var color)) img.color = color;

            string type = spec["type"]?.ToString();
            if (!string.IsNullOrEmpty(type) && Enum.TryParse<Image.Type>(type, true, out var imgType))
                img.type = imgType;

            if (spec["raycast_target"] != null)
            {
                try { img.raycastTarget = spec["raycast_target"].ToObject<bool>(); } catch { }
            }
            if (spec["preserve_aspect"] != null)
            {
                try { img.preserveAspect = spec["preserve_aspect"].ToObject<bool>(); } catch { }
            }
        }

        private static void ConfigureText(GameObject go, JObject spec)
        {
            var tmpType = TmpTextType;
            if (tmpType == null)
            {
                McpLog.Warn($"[ManageCanvas] TextMeshPro (Unity.TextMeshPro) not available — text node '{go.name}' skipped.");
                return;
            }

            var tmp = go.GetComponent(tmpType);
            if (tmp == null) tmp = go.AddComponent(tmpType);

            SetReflectProp(tmp, "text", spec["text"]?.ToString());
            if (spec["font_size"] != null)
            {
                try { SetReflectProp(tmp, "fontSize", spec["font_size"].ToObject<float>()); } catch { }
            }
            if (TryReadColor(spec["color"], out var color)) SetReflectProp(tmp, "color", color);

            string alignment = spec["alignment"]?.ToString();
            if (!string.IsNullOrEmpty(alignment))
            {
                var alignType = Type.GetType("TMPro.TextAlignmentOptions, Unity.TextMeshPro");
                if (alignType != null)
                {
                    try { SetReflectProp(tmp, "alignment", Enum.Parse(alignType, alignment, true)); } catch { }
                }
            }

            string fontStyle = spec["font_style"]?.ToString();
            if (!string.IsNullOrEmpty(fontStyle))
            {
                var styleType = Type.GetType("TMPro.FontStyles, Unity.TextMeshPro");
                if (styleType != null)
                {
                    try { SetReflectProp(tmp, "fontStyle", Enum.Parse(styleType, fontStyle, true)); } catch { }
                }
            }

            string fontAsset = spec["font_asset"]?.ToString();
            if (!string.IsNullOrEmpty(fontAsset))
            {
                var asset = AssetDatabase.LoadAssetAtPath<UnityEngine.Object>(fontAsset);
                if (asset != null) SetReflectProp(tmp, "font", asset);
            }

            if (spec["raycast_target"] != null)
            {
                try { SetReflectProp(tmp, "raycastTarget", spec["raycast_target"].ToObject<bool>()); } catch { }
            }
        }

        private static GameObject ResolveParent(string parentRef, out string error)
        {
            error = null;
            if (string.IsNullOrEmpty(parentRef)) return null;

            if (int.TryParse(parentRef, out int id))
            {
                var obj = EditorUtility.InstanceIDToObject(id) as GameObject;
                if (obj == null) error = $"No GameObject with instance ID {id}.";
                return obj;
            }

            var found = GameObject.Find(parentRef);
            if (found == null) error = $"Parent GameObject not found: '{parentRef}'.";
            return found;
        }

        private static string GetGameObjectPath(GameObject go)
        {
            if (go == null) return null;
            var sb = new StringBuilder(go.name);
            var t = go.transform.parent;
            while (t != null)
            {
                sb.Insert(0, t.name + "/");
                t = t.parent;
            }
            return sb.ToString();
        }

        private static bool TryReadVector2(JToken token, out Vector2 v)
        {
            v = Vector2.zero;
            if (token is JArray arr && arr.Count >= 2)
            {
                try
                {
                    v = new Vector2(arr[0].ToObject<float>(), arr[1].ToObject<float>());
                    return true;
                }
                catch { }
            }
            return false;
        }

        private static bool TryReadColor(JToken token, out Color c)
        {
            c = Color.white;
            if (token is JArray arr && arr.Count >= 3)
            {
                try
                {
                    float a = arr.Count >= 4 ? arr[3].ToObject<float>() : 1f;
                    c = new Color(arr[0].ToObject<float>(), arr[1].ToObject<float>(), arr[2].ToObject<float>(), a);
                    return true;
                }
                catch { }
            }
            return false;
        }

        private static void SetReflectProp(object obj, string propName, object value)
        {
            if (obj == null || value == null) return;
            var prop = obj.GetType().GetProperty(propName);
            if (prop == null || !prop.CanWrite) return;
            try { prop.SetValue(obj, value); }
            catch { /* type mismatch — ignore so partial spec still applies */ }
        }
    }
}

File: Server/src/services/tools/manage_canvas.py
"""
Defines the manage_canvas tool for building runtime uGUI hierarchies
(Canvas / RectTransform / Image / TextMeshProUGUI) from a declarative spec.

Distinct from manage_ui, which targets UI Toolkit (UXML/USS/VisualElement).
"""
from typing import Annotated, Any, Literal

from fastmcp import Context
from mcp.types import ToolAnnotations

from services.registry import mcp_for_unity_tool
from services.tools import get_unity_instance_from_context
from services.tools.utils import parse_json_payload
from transport.unity_transport import send_with_unity_instance
from transport.legacy.unity_connection import async_send_command_with_retry


@mcp_for_unity_tool(
    group="ui",
    description=(
        "Builds a runtime uGUI hierarchy (Canvas / RectTransform / Image / TextMeshProUGUI) from a "
        "declarative spec in one call. Use this instead of stitching together manage_gameobject + "
        "manage_components calls for layered UI like a plate-with-icon-and-text or a toolbar with cells. "
        "Distinct from manage_ui (UI Toolkit / UXML / USS).\n\n"
        "Action: build.\n\n"
        "Layout spec (nested dict):\n"
        "  name (required): GameObject name.\n"
        "  rect: dict with optional keys:\n"
        "    anchor: anchor preset string. One of: top-left, top-center, top-right, "
        "middle-left, middle-center (or 'center'), middle-right, bottom-left, bottom-center, "
        "bottom-right, stretch-top, stretch-bottom, stretch-left, stretch-right, "
        "stretch-horizontal, stretch-vertical, stretch (or 'stretch-all').\n"
        "    anchor_min, anchor_max, pivot: [x, y] 0-1 floats (override preset when present).\n"
        "    anchored_position: [x, y] in pixels.\n"
        "    size_delta: [w, h] in pixels (or [horiz_pad, vert_pad] when an axis is stretched).\n"
        "    rotation: z-axis euler degrees.\n"
        "    scale: [x, y] (z stays 1).\n"
        "  image: dict, adds a UnityEngine.UI.Image:\n"
        "    sprite: Assets-relative path to a Sprite.\n"
        "    color: [r, g, b] or [r, g, b, a], 0-1 floats.\n"
        "    type: 'Simple' | 'Sliced' | 'Tiled' | 'Filled'.\n"
        "    raycast_target (bool), preserve_aspect (bool).\n"
        "  text: dict, adds a TMPro.TextMeshProUGUI (requires the TextMeshPro package):\n"
        "    text: string content.\n"
        "    font_size: number.\n"
        "    color: [r, g, b, a] 0-1.\n"
        "    alignment: TextAlignmentOptions name (Center, Left, Right, MidlineCenter, etc.).\n"
        "    font_style: FontStyles name (Normal, Bold, Italic, ...).\n"
        "    font_asset: Assets-relative path to a TMP_FontAsset.\n"
        "    raycast_target (bool).\n"
        "  children: list of nested layout specs.\n\n"
        "Returns: root {name, instanceID, path} plus a flat 'nodes' list of every created GameObject."
    ),
    annotations=ToolAnnotations(
        title="Manage Canvas (uGUI Builder)",
        destructiveHint=True,
    ),
)
async def manage_canvas(
    ctx: Context,
    action: Annotated[Literal["build"], "Action to perform."],

    layout: Annotated[dict[str, Any] | str,
                      "Nested layout spec dict (see tool description). Accepts a JSON string for clients "
                      "that cannot send nested objects directly."] | None = None,

    parent: Annotated[str,
                      "Where to attach the root. GameObject name, hierarchy path (e.g. 'Canvas/HUD'), or "
                      "instance ID as string. Omit to attach to the scene root."] | None = None,
) -> dict[str, Any]:
    unity_instance = await get_unity_instance_from_context(ctx)

    if action.lower() != "build":
        return {"success": False, "message": f"Unknown action '{action}'. Valid: build."}

    if isinstance(layout, str):
        parsed = parse_json_payload(layout)
        if not isinstance(parsed, dict):
            return {"success": False, "message": "'layout' string must parse to a JSON object."}
        layout = parsed

    if not isinstance(layout, dict):
        return {"success": False, "message": "'layout' is required and must be an object."}

    if not layout.get("name") and not layout.get("children"):
        return {"success": False, "message": "'layout' must have a 'name' (or at minimum 'children')."}

    params_dict: dict[str, Any] = {
        "action": "build",
        "layout": layout,
    }
    if parent is not None:
        params_dict["parent"] = parent

    result = await send_with_unity_instance(
        async_send_command_with_retry,
        unity_instance,
        "manage_canvas",
        params_dict,
    )

    return result if isinstance(result, dict) else {"success": False, "message": str(result)}
